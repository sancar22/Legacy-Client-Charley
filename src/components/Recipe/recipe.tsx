import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { BsPencil } from 'react-icons/bs';
import { IRecipe, IState } from 'src/interfaces';
import { delete_item, add_item } from '../../state/actions';
import apiService from '../../services/apiService';
import RecipeModal from '../Modals/RecipeModal/recipeModal';
import EditModal from '../Modals/EditModal/editModal';
import * as styles from './recipe.module.css';

const Recipe = ({
  recipe,
  remove,
  edit,
  save,
  self,
}: {
  recipe: IRecipe;
  remove: boolean;
  edit: boolean;
  save: boolean;
  self: boolean;
}): JSX.Element => {
  const [inFocus, setInFocus] = useState(false);
  const [modalStatus, setModalStatus] = useState(false);
  const [editModalStatus, setEditModalStatus] = useState(false);
  const [saved, setSaved] = useState(false);
  const { username } = useSelector<IState>((state) => state.username);

  const dispatch = useDispatch();

  const handleModal = () => {
    setModalStatus(!modalStatus);
  };
  const handleEditModal = () => {
    setEditModalStatus(!editModalStatus);
  };

  const handleDelete = async () => {
    try {
      await apiService.deleteRecipe(recipe._id);
      dispatch(delete_item(recipe._id));
      setSaved(false);
    } catch (e) {
      console.log(e);
    }
  };

  const handleSave = async () => {
    try {
      await apiService.addFromFriend(recipe);
      dispatch(add_item(recipe));
      setSaved(true);
    } catch (e) {
      console.log(e);
    }
  };

  let keywordString = '';
  if (recipe.keywords) {
    keywordString = recipe.keywords.join(', ');
  }

  return (
    <>
      <div
        className={styles.container}
        onMouseEnter={() => setInFocus(true)}
        onMouseLeave={() => setInFocus(false)}
        aria-hidden='true'
      >
        {recipe.image && (
          <img src={recipe.image} className={styles.image} alt='recipe' />
        )}
        {!recipe.image && <div className={styles.noImage}></div>}

        <div
          className={styles.detailsBox}
          onClick={handleModal}
          aria-hidden='true'
        >
          <div>
            <div className={styles.details__name}>{recipe.name}</div>
            {self ? (
              <>
                {recipe.origin === username ? null : (
                  <div className={styles.details__origin}>
                    {`from chef ${recipe.origin}`}
                  </div>
                )}
              </>
            ) : null}
            <div className={styles.details__author}>{recipe.publisher}</div>
            <div className={styles.details__author}>{recipe.author}</div>
          </div>
          <div className={styles.details__keywords}>{keywordString}</div>
        </div>

        <div className={styles.buttons}>
          {remove && (
            <div
              className={inFocus ? styles.button__show : styles.button__hide}
              onClick={handleDelete}
              aria-hidden='true'
            >
              x
            </div>
          )}
          {edit && (
            <div
              className={
                inFocus ? styles.editButton__show : styles.editButton__hide
              }
              onClick={handleEditModal}
              aria-hidden='true'
            >
              <BsPencil />
            </div>
          )}
          {save && (
            <div
              className={inFocus ? styles.button__show : styles.button__hide}
              onClick={handleSave}
              aria-hidden='true'
            >
              +
            </div>
          )}
        </div>
      </div>
      <RecipeModal
        show={modalStatus}
        handleClose={handleModal}
        recipe={recipe}
      />
      <EditModal
        show={editModalStatus}
        handleClose={handleEditModal}
        recipe={recipe}
      />
    </>
  );
};

export default Recipe;
