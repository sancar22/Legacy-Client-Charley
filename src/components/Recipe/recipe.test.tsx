import React from 'react';
import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import reduxStore from '../../state/store';
import Recipe from './recipe';

const { store } = reduxStore();

describe('Recipe component', () => {
  const recipe = create(
    <Provider store={store}>
      <Recipe recipe = {{
        _id: '',
        notes: [],
        origin: '',
        url: '',
        recipeIngredient: [],
        recipeInstructions: [],
      }} />
    </Provider>,
  );
  test('match previous Recipe snapshot', () => {
    expect(recipe).toMatchSnapshot();
  });
});
