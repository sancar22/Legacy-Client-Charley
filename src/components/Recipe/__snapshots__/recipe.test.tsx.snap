// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe component match previous Recipe snapshot 1`] = `
Array [
  <div
    aria-hidden="true"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div />
    <div
      aria-hidden="true"
      onClick={[Function]}
    >
      <div>
        <div />
        <div />
        <div />
      </div>
      <div>
        
      </div>
    </div>
    <div />
  </div>,
  <div>
    <div
      aria-hidden="true"
      onClick={[Function]}
    >
      x
    </div>
    <div />
    <div>
      <div />
      <ul />
      <div>
        Prepare
      </div>
    </div>
  </div>,
  <div>
    <form
      onSubmit={[Function]}
    >
      <div>
        <div
          aria-hidden="true"
          onClick={[Function]}
        >
          x
        </div>
        <button
          type="submit"
        >
          save
        </button>
      </div>
      <input
        onChange={[Function]}
      />
    </form>
    <div>
      Notes
    </div>
    <div
      aria-hidden="true"
      onClick={[Function]}
    >
      edit your notes!
    </div>
    <ul />
  </div>,
]
`;
